src/events/broker.py

```diff
@@ -482,8 +495,102 @@ def _format_storages(self) -> None:
         )
 
+    def _leader_elected(self, event: LeaderElectedEvent) -> None:
+        if self.charm.state.runs_controller and self.charm.state.cluster.bootstrap_controller:
+
+            updated_bootstrap_data = {
+                "bootstrap-controller": self.charm.state.bootstrap_controller,
+                "bootstrap-unit-id": str(self.charm.state.kraft_unit_id),
+                "bootstrap-replica-id": self.charm.state.unit_broker.directory_id,
+            }
+            self.charm.state.cluster.update(updated_bootstrap_data)
+
+            if self.charm.state.peer_cluster_orchestrator:
+                self.charm.state.peer_cluster_orchestrator.update(updated_bootstrap_data)
+
+            # A rolling restart is required to apply the new leader config
+            self.charm.on[f"{self.charm.restart.name}"].acquire_lock.emit()
+
+    @retry(
+        wait=wait_fixed(15),
+        stop=stop_after_attempt(4),
+        reraise=True,
+    )
+    def _add_controller(self) -> None:
+        """Adds current unit to the dynamic quorum in KRaft mode if this is a follower unit."""
+        if (
+            self.charm.state.runs_controller
+            and not self.charm.unit.is_leader()
+            and not self.charm.state.unit_broker.added_to_quorum
+        ):
+            result = self.workload.run_bin_command(
+                bin_keyword="metadata-quorum",
+                bin_args=[
+                    "--bootstrap-controller",
+                    self.charm.state.cluster.bootstrap_controller,
+                    "--command-config",
+                    self.workload.paths.server_properties,
+                    "add-controller",
+                ],
+            )
+            logger.debug(result)
+            directory_id = self.workload.get_directory_id(self.charm.state.log_dirs)
+            self.charm.state.unit_broker.update(
+                {"directory-id": directory_id, "added-to-quorum": "true"}
+            )
+
+    @retry(
+        wait=wait_fixed(10),
+        stop=stop_after_attempt(3),
+        reraise=True,
+    )
+    def _remove_controller(
+        self, controller_id: int, controller_directory_id: str, bootstrap_node: str | None = None
+    ):
+        if not bootstrap_node:
+            bootstrap_node = self.charm.state.cluster.bootstrap_controller
+
+        try:
+            self.workload.run_bin_command(
+                bin_keyword="metadata-quorum",
+                bin_args=[
+                    "--bootstrap-controller",
+                    bootstrap_node,
+                    "--command-config",
+                    self.workload.paths.server_properties,
+                    "remove-controller",
+                    "--controller-id",
+                    str(controller_id),
+                    "--controller-directory-id",
+                    controller_directory_id,
+                ],
+            )
+        except Exception as e:
+            error_details = getattr(e, "stderr")
```

Comment:
**question:** Why `getattr` instead of `e.stderr`? I would have imagined you would default to an empty string if there wasn't that attr.

Reply:
Honestly, I was avoiding an import here :) now fixed.